pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	managedImages := make(map[string]bool)
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+		if !t.repo.pullthrough {
+			managed, found := managedImages[history.Items[0].Image]
+			if !found {
+				dgst, err := digest.ParseDigest(history.Items[0].Image)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+					continue
+				}
+
+				image, err := t.repo.getImage(dgst)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
+					continue
+				}
+				managed = isImageManaged(image)
+				managedImages[history.Items[0].Image] = managed
+			}
+
+			if !managed {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	managedImages := make(map[string]bool)
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+			continue
+		}
+
+		if dgst != desc.Digest {
+			continue
+		}
+
+		if !t.repo.pullthrough {
+			managed, found := managedImages[history.Items[0].Image]
+			if !found {
+				image, err := t.repo.getImage(dgst)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
+					continue
+				}
+				managed = isImageManaged(image)
+				managedImages[history.Items[0].Image] = managed
+			}
+
+			if !managed {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+
+	return tags, nil
+}
+
+func (t tagService) Tag(ctx context.Context, tag string, dgst distribution.Descriptor) error {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	image, err := t.repo.registryOSClient.Images().Get(dgst.Digest.String())
+	if err != nil {
+		context.GetLogger(ctx).Errorf("unable to get image: %s", dgst.Digest.String())
+		return err
+	}
+	image.SetResourceVersion("")
+
+	if !t.repo.pullthrough && !isImageManaged(image) {
```

Comment:
Who calls the tag operation?  The push code?  Or is this available directly to end clients (i.e. a client could say "tag X to digest Y")?  If it's the latter, then all the code after this if check is a security hole (would allow users to tag arbitrary digests).  If it's the former, then whether the image is managed or not does not matter.